src/main/java/org/broadinstitute/hellbender/tools/spark/validation/BqsrRecalBasesCompare.java

```diff
@@ -0,0 +1,230 @@
+        JavaRDD<GATKRead> secondReads = readsSource2.getParallelReads(input2, getIntervals());
+
+        long firstBamSize = firstReads.count();
+        long secondBamSize = secondReads.count();
+        if (firstBamSize != secondBamSize) {
+            throw new UserException("input bams have different numbers of total reads: "
+                    + firstBamSize + "," + secondBamSize);
+        }
+        System.out.println("Processing bams with " + firstBamSize + " reads");
+
+        // To compare bases, we created two PairRDDs keyed by read name, strand, etc. where the value is the bases.
+        // We cogroup the two RDDs, verify there is only one read per key per bam, and then diff the bases
+        // position by position for the two reads and put the results in CompareMatrix.
+        JavaPairRDD<String, Quals> firstQuals = firstReads.mapToPair(read -> {
+            read.clearAttributes();
+            return new Tuple2<>(read.toString(), new Quals(read.getBaseQualities()));
+        });
+        JavaPairRDD<String, Quals> secondQuals = secondReads.mapToPair(read -> {
+            read.clearAttributes();
+            return new Tuple2<>(read.toString(), new Quals(read.getBaseQualities()));
+        });
+
+        JavaPairRDD<String, Tuple2<Iterable<Quals>, Iterable<Quals>>> cogroup = firstQuals.cogroup(secondQuals);
+        CompareMatrix finalMatrix = cogroup.map(v1 -> {
+            List<Quals> firstQuals1 = Lists.newArrayList(v1._2()._1());
+            List<Quals> secondQuals1 = Lists.newArrayList(v1._2()._2());
+            CompareMatrix compareMatrix = new CompareMatrix();
+            if (firstQuals1.size() != 1) {
+                throw new GATKException("expected only one read per key in first bam: " + firstQuals1.size());
+            }
+            if (secondQuals1.size() != 1) {
+                throw new GATKException("expected only one read per key in second bam: " + secondQuals1.size());
+            }
+
+            compareMatrix.add(firstQuals1.get(0).quals, secondQuals1.get(0).quals);
+            return compareMatrix;
+        }).treeAggregate(new CompareMatrix(),
+                (Function2<CompareMatrix, CompareMatrix, CompareMatrix>) CompareMatrix::addMatrix,
+                (Function2<CompareMatrix, CompareMatrix, CompareMatrix>) CompareMatrix::addMatrix);
+
+
+        if (printSummary) {
+            finalMatrix.printSummary();
+        }
+
+        if (printMatrix) {
+            finalMatrix.print();
+        }
+
+        if (crashOnDiff && finalMatrix.hasNondiaginalElements()) {
+            throw new UserException("Quality scores from the two BAMs do not match");
+        }
+    }
+
+}
+
+/**
+ * CompareMatrix contains a square matrix of linear dimension QualityUtils.MAX_SAM_QUAL_SCORE.
+ * The elements are counts of the number of times bam 1 had a quality score of i and bam 2 had a quality
+ * score of j, (stored in mat[i][j]).
+ */
+class CompareMatrix implements Serializable {
```

Comment:
Couldn't we just call `Arrays.equals()` on the two quals arrays? Can you explain why this more complex approach is needed?